app/ts/store/provider.ts

```diff
@@ -0,0 +1,45 @@
+import { effect, signal } from '@preact/signals'
+import { ethers } from 'ethers'
+import { assertsWithEthereum } from '../library/ethereum.js'
+import { WalletError } from '../library/exceptions.js'
+import { Web3Provider } from '../types.js'
+
+const provider = signal<Web3Provider | undefined>(undefined)
+
+export function useProviders() {
+	const getbrowserProvider = () => {
+		if (provider.value !== undefined) return provider.value
+
+		try {
+			assertsWithEthereum(window)
+			provider.value = new ethers.providers.Web3Provider(window.ethereum)
+			return provider.value
+		} catch (error) {
+			if (error instanceof WalletError) {
+				throw error
+			}
+
+			const errorMessage = typeof error === 'string' ? error : error instanceof Error ? error.message : 'An unknown error occurred.'
+			throw new Error(errorMessage)
+		}
+	}
+
+	return {
+		browserProvider: provider,
+		getbrowserProvider,
+	}
+}
+
+const handleChainChange = async () => {
+	removeChainChangeListener()
+
+	// reinitialize provider
+	assertsWithEthereum(window)
+	provider.value = new ethers.providers.Web3Provider(window.ethereum)
+}
+
+const removeChainChangeListener = effect(() => {
+	if (provider.value === undefined) return
+	assertsWithEthereum(window)
+	window.ethereum.on('chainChanged', handleChainChange)
+})
```

Comment:
When network changes you don't need to reinitialize the provider, for a future optimization you could just make sure to remove event listener on disconnect instead.
Right now you, set the value of the provider store again, which triggers Signal effect in `app/ts/store/network.ts` which then queries the network from the provider.
You could just use the network provided in the network change event, this would remove the need for a signalEffect and rpc calls to fetch network.

Reply:
Will take a look closer as a separate issue #106